File: android/src/main/java/com/sendbird/calls/reactnative/module/CallsQueries.kt
package com.sendbird.calls.reactnative.module

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableNativeArray
import com.sendbird.calls.*
import com.sendbird.calls.reactnative.RNCallsInternalError
import com.sendbird.calls.reactnative.extension.rejectCalls
import com.sendbird.calls.reactnative.module.listener.CallsGroupCallListener
import com.sendbird.calls.reactnative.utils.CallsUtils

enum class QueryType {
    DIRECT_CALL_LOG,
    ROOM_LIST
}

class CallsQueries(private val root: CallsModule) {
    val directCallLogQueries = mutableMapOf<String, DirectCallLogListQuery>()
    val roomListQueries = mutableMapOf<String, RoomListQuery>()

    fun createDirectCallLogListQuery(params: ReadableMap, promise: Promise) {
        val queryParams = DirectCallLogListQuery.Params().apply {
            CallsUtils.safeGet {
                params.getInt("limit")
            }?.let { setLimit(it) }

            CallsUtils.safeGet {
                params.getString("myRole")?.let { DirectCallUserRole.valueOf(it) }
            }?.let { setMyRole(it) }

            CallsUtils.safeGet {
                params.getArray("endResults")?.toArrayList()?.map { DirectCallEndResult.valueOf(it as String) }
            }?.let { setEndResults(it) }
        }

        SendBirdCall.createDirectCallLogListQuery(queryParams).let {
            it.hashCode().toString().run {
                directCallLogQueries[this] = it
                promise.resolve(this)
            }
        }
    }

    fun createRoomListQuery(params: ReadableMap, promise: Promise) {
        val queryParams = RoomListQuery.Params().apply {
            CallsUtils.safeGet {
                params.getInt("limit")
            }?.let { setLimit(it) }

            CallsUtils.safeGet {
                params.getArray("createdByUserIds")?.toArrayList()?.map { it as String }
            }?.let { setCreatedByUserIds(it) }

            CallsUtils.safeGet {
                params.getArray("roomIds")?.toArrayList()?.map { it as String }
            }?.let { setRoomIds(it) }

            CallsUtils.safeGet {
                params.getString("state")?.let { RoomState.valueOf(it) }
            }?.let { setState(it) }

            CallsUtils.safeGet {
                params.getString("type")?.let { RoomType.valueOf(it) }
            }?.let { setType(it) }


            // FIXME: Range methods are internal. (reported issue)
//            val createdAtRange = CallsUtils.safeGet {
//                val range = Range()
//                val createdAt = params.getMap("createdAt") ?: WritableNativeMap()
//                CallsUtils
//                    .safeGet { createdAt.getDouble("lowerBound") }
//                    ?.let { range.greaterThanOrEqualTo(it.toLong()) }
//                CallsUtils
//                    .safeGet { createdAt.getDouble("upperBound") }
//                    ?.let { range.lessThanOrEqualTo(it.toLong()) }
//                range
//            }
//
//            val participantCountRange = CallsUtils.safeGet {
//                val range = Range()
//                val count = params.getMap("currentParticipantCount")
//                CallsUtils
//                    .safeGet { count?.getDouble("lowerBound") }
//                    ?.let { range.greaterThanOrEqualTo(it.toLong()) }
//                CallsUtils
//                    .safeGet { count?.getDouble("upperBound") }
//                    ?.let { range.lessThanOrEqualTo(it.toLong()) }
//                range
//            }
        }

        SendBirdCall.createRoomListQuery(queryParams).let {
            it.hashCode().toString().run {
                roomListQueries[this] = it
                promise.resolve(this)
            }
        }
    }

    fun queryNext(key: String, type: String, promise: Promise) {
        when (QueryType.valueOf(type)) {
            QueryType.DIRECT_CALL_LOG -> {
                directCallLogQueries[key]
                    ?.run {
                        this.next { list, error ->
                            error?.let {
                                promise.rejectCalls(error)
                            }
                            list?.let {
                                promise.resolve(CallsUtils.convertToJsMap(mapOf(
                                    "hasNext" to this.hasNext(),
                                    "result" to list.map {
                                        CallsUtils.convertDirectCallLogToJsMap(it)
                                    },
                                )))
                            }
                        }
                    }
                    ?: run {
                        promise.reject(RNCallsInternalError("queryNext", RNCallsInternalError.Type.NOT_FOUND_QUERY))
                    }
            }
            QueryType.ROOM_LIST -> {
                roomListQueries[key]
                    ?.run {
                        this.next { list, error ->
                            error
                                ?.let {
                                    promise.rejectCalls(it)
                                }
                                ?: run {
                                    promise.resolve(CallsUtils.convertToJsMap(mapOf(
                                        "hasNext" to this.hasNext(),
                                        "result" to list.map {
                                            it.addListener(it.roomId, CallsGroupCallListener.get(root, it))
                                            CallsUtils.convertRoomToJsMap(it)
                                        }
                                    )))
                                }
                        }
                    }
                    ?: run {
                        promise.reject(RNCallsInternalError("queryNext", RNCallsInternalError.Type.NOT_FOUND_QUERY))
                    }
            }
            else -> {
                promise.resolve(CallsUtils.convertToJsMap(mapOf(
                    "hasNext" to false,
                    "result" to WritableNativeArray()
                )))
            }
        }
    }
    fun queryRelease(key: String) {
        directCallLogQueries.remove(key)
        roomListQueries.remove(key)
    }
}

File: ios/Modules/CallsQueries.swift
//
//  CallsQueries.swift
//  RNSendbirdCalls
//
//  Created by Airen Kang on 2022/06/15.
//  Copyright © 2022 Facebook. All rights reserved.
//

import Foundation
import SendBirdCalls

enum QueryType {
    case directCallLog
    case roomList
    
    var asString: String {
        switch self {
        case .directCallLog:
            return "DIRECT_CALL_LOG"
        case .roomList:
            return "ROOM_LIST"
        }
    }
    
    init?(fromString: String) {
        switch fromString {
        case "DIRECT_CALL_LOG":
            self = .directCallLog
        case "ROOM_LIST":
            self = .roomList
        default:
            return nil
        }
    }
}

class CallsQueries: NSObject {
    var directCallLogQueries: [String: DirectCallLogListQuery] = [:]
    var roomListQueries: [String: RoomListQuery] = [:]
    
    func createDirectCallLogListQuery(_ params: [String: Any], _ promise: Promise) {
        let queryParams = DirectCallLogListQuery.Params()
        
        if let myRole = params["myRole"] as? String,
           let myRoleFilter = DirectCallLogListQuery.UserRoleFilter.init(fromString: myRole)  {
            queryParams.myRole = myRoleFilter
        }
        
        if let limit = params["limit"] as? NSNumber {
            queryParams.limit = Int32(truncating: limit)
        }
        
        if let endResults = params["endResults"] as? Array<String> {
            queryParams.endResults = endResults.map { DirectCallEndResult.init(fromString: $0) }
        }
        
        guard let query = SendBirdCall.createDirectCallLogListQuery(with: queryParams) else {
            return promise.reject(.queryCreateFailure("createDirectCallLogListQuery"))
        }
        
        let key = query.hash.description
        directCallLogQueries[key] = query
        promise.resolve(key)
    }
    
    func createRoomListQuery(_ params: [String: Any], _ promise: Promise) {
        let queryParams = RoomListQuery.Params()
        
        if let limit = params["limit"] as? NSNumber {
            queryParams.setLimit(UInt(truncating: limit))
        }
        
        if let createdByUserIds = params["createdByUserIds"] as? Array<String> {
            queryParams.setCreatedByUserIds(createdByUserIds)
        }
        
        if let roomIds = params["roomIds"] as? Array<String> {
            queryParams.setRoomIds(roomIds)
        }
        
        if let state = params["state"] as? String,
           let roomState = SendBirdCalls.Room.State(fromString: state) {
            queryParams.setState(roomState)
        }
        
        if let type = params["type"] as? String,
           let roomType = SendBirdCalls.RoomType(fromString: type) {
            queryParams.setType(roomType)
        }
        
        if let createdAt = params["createdAt"] as? [String: Any] {
            let range = SendBirdRange()
            if let lowerBound = createdAt["lowerBound"] as? NSNumber {
                range.lowerBound = Int64(truncating: lowerBound)
            }
            if let upperBound = createdAt["upperBound"] as? NSNumber {
                range.upperBound = Int64(truncating: upperBound)
            }
            queryParams.createdAtRange = range
        }
        
        if let count = params["currentParticipantCount"] as? [String: Any] {
            let range = SendBirdRange()
            if let lowerBound = count["lowerBound"] as? NSNumber {
                range.lowerBound = Int64(truncating: lowerBound)
            }
            if let upperBound = count["upperBound"] as? NSNumber {
                range.upperBound = Int64(truncating: upperBound)
            }
            queryParams.currentParticipantCountRange = range
        }
        
        guard let query = SendBirdCall.createRoomListQuery(with: queryParams) else {
            return promise.reject(.queryCreateFailure("createRoomListQuery"))
        }
        
        let key = query.hash.description
        roomListQueries[key] = query
        promise.resolve(key)
    }
    
    func queryNext(_ key: String, _ type: String, _ promise: Promise) {
        guard let type = QueryType(fromString: type) else {
            return promise.resolve([ "hasNext": false,"result": [] ])
        }
        
        switch type {
        case .directCallLog:
            guard let query = directCallLogQueries[key] else { return }
            query.next { list, error in
                if let error = error {
                    promise.reject(error)
                }
                
                if let list = list {
                    promise.resolve([
                        "hasNext": query.hasNext,
                        "result": list.map {
                            CallsUtils.convertDirectCallLogToDict($0)
                        }
                    ])
                }
            }
            
        case .roomList:
            guard let query = roomListQueries[key] else { return }
            query.next { list, error in
                if let error = error {
                    promise.reject(error)
                }
                
                if let list = list {
                    promise.resolve([
                        "hasNext": query.hasNext,
                        "result": list.map {
                            $0.addDelegate(GroupCallDelegate.get($0), identifier: $0.roomId)
                            CallsUtils.convertRoomToDict($0)
                        }
                    ])
                }
            }
        }
    }
    
    func queryRelease(_ key: String) {
        directCallLogQueries.removeValue(forKey: key)
        roomListQueries.removeValue(forKey: key)
    }
}
